### DISCUSSION TITLE
How to debug OpenThread?

### DISCUSSION THREAD

**Discussion by AuSchell** [openthread/openthread#8949] [Q&A] [answered]
Hello everyone,

I would like to debug my OpenThread software, but building a degug instance seems not to be implemented at this [point](https://github.com/openthread/openthread/tree/main/examples). Is there a workaround to build a debug instance? Will the feature be implemented in the future?

**Comment by Irving-cl** [comment, marked as answer]
Hi @AuSchell,

Do you mean to have the '-g' option of gcc when building the instance? I'm not quite familiar with autotools. I suggest using cmake to build the code. If using cmake, I think you will have a debug build by default. See the top-level [CMakeLists.txt](https://github.com/openthread/openthread/blob/main/CMakeLists.txt):
```
if(NOT CMAKE_BUILD_TYPE)
    # Check if this is a top-level CMake.
    # If it is not, do not set the CMAKE_BUILD_TYPE because OpenThread is a part of something bigger.
    if ("${CMAKE_PROJECT_NAME}" STREQUAL "openthread")
        set(CMAKE_BUILD_TYPE "Debug" CACHE STRING "default build type: Debug" FORCE)
    endif ()
endif()
```

To build with CMake, there are some scripts provided in OT. Not sure what targets you want to build. For example, if you want to build `ot-cli-xxx` on simulation platform, you can build it using:
```
./script/cmake-build simulation -DOT_APP_NCP=OFF
```

**Reply by AuSchell** [reply]
Hi @Irving-cl,

Thank you for your answer. It put me on the right track, the option "-DCMAKE_BUILD_TYPE" must have the value "Debug". In my case I build for the nrf52840, so I had to change the value in the [build script](https://github.com/openthread/ot-nrf528xx/blob/974903d95786a2b88fc9f6ac7478070456624bb8/script/build#L44) from "MinSizeRel" to "Debug".
